### PR TITLE
Send 500 and close connection if file operations fail

### PR DESCRIPTION
**Pull request by deadtrickster**
The idea is that Elli was asked to send a file that doesn't exist or can't be accessed we send 500.

Also send bad request body (Content-Length was set but no body)

### CODE REVIEW

**Comment by deadtrickster** [comment]
ping

**Comment by yurrriq** [comment]
Nice, it's green! I'll have a look a little later.

File: src/elli_example_callback.erl [hunk 1]
```diff
@@ -135,10 +135,15 @@ handle('GET', [<<"decoded-list">>], Req) ->
 handle('GET', [<<"sendfile">>], _Req) ->
     %% Returning {file, "/path/to/file"} instead of the body results
     %% in Elli using sendfile.
-    %% All required headers should be added by the handler.
```
Comment (yurrriq) [inline src/elli_example_callback.erl:138]: Is there a particular reason to remove this comment?

File: src/elli_http.erl [hunk 2]
```diff
@@ -87,8 +87,8 @@ keepalive_loop(Socket, NumRequests, Buffer, Options, Callback) ->
       Callback  :: elli_handler:callback(),
       ConnToken :: {'keep_alive' | 'close', binary()}.
 handle_request(S, PrevB, Opts, {Mod, Args} = Callback) ->
-    {Method, RawPath, V, B0} = get_request(S, PrevB, Opts, Callback),
     t(request_start),
+    {Method, RawPath, V, B0} = get_request(S, PrevB, Opts, Callback),
```
Comment (yurrriq) [inline src/elli_http.erl:91]: Is this the right place, after `t(request_start)`?

File: src/elli_http.erl [hunk 3]
```diff
@@ -167,38 +167,46 @@ handle_response(Req, Buffer, {file, ResponseCode, UserHeaders,
...
-                 Args),
+    case elli_util:file_size(Filename) of
+        {error, FileError} ->
+            handle_event(Mod, file_error, [FileError], Args),
+            send_server_error(Req#req.socket),
+            elli_tcp:close(Req#req.socket),
+            exit(normal);
+        Size ->
+            t(send_start),
+            case elli_util:normalize_range(Range, Size) of
```
Comment (yurrriq) [inline src/elli_http.erl:178]: In a future release, we should try to pull these nested `case` expressions out to (a) helper function(s), but I think this is fine for now.

File: src/elli_http.erl [hunk 4]
```diff
@@ -262,10 +274,16 @@ do_send_file(Fd, {Offset, Length}, #req{callback={Mod, Args}} = Req, Headers) ->
...
 send_bad_request(Socket) ->
-    Body = <<"Bad Request">>,
-    Response = [<<"HTTP/1.1 ">>, status(400), <<"\r\n">>,
+    send_rescue_response(Socket, 400, <<"Bad Request">>).
+
+send_server_error(Socket) ->
+    send_rescue_response(Socket, 500, <<"Server Error">>).
+
+send_rescue_response(Socket, Code, Body) ->
+    Response = [<<"HTTP/1.1 ">>, status(Code), <<"\r\n">>,
```
Comment (yurrriq) [inline src/elli_http.erl:283]: `1.1` should not be hard-coded here. See [`#req.version`](https://github.com/elli-lib/elli/blob/7e9cadfe3728bcada360c606c869d3361faf4aa3/include/elli.hrl#L5) and [`elli_http:version/0`](https://github.com/elli-lib/elli/blob/7e9cadfe3728bcada360c606c869d3361faf4aa3/src/elli_http.erl#L25).

File: src/elli_http.erl [hunk 5]
```diff
@@ -262,10 +274,16 @@ do_send_file(Fd, {Offset, Length}, #req{callback={Mod, Args}} = Req, Headers) ->
...
                 <<"Content-Length: ">>, integer_to_list(size(Body)), <<"\r\n">>,
-                <<"\r\n">>],
+                <<"\r\n">>,
+                Body],
```
Comment (yurrriq) [inline src/elli_http.erl:286]: Why send a body here at all?

**Review by yurrriq — CHANGES_REQUESTED** [review]

File: src/elli_example_callback.erl (hunk 1 above)
Comment (deadtrickster) [reply src/elli_example_callback.erl:138]: Content-Length is required but set automatically

File: src/elli_http.erl (hunk 2 above)
Comment (deadtrickster) [reply src/elli_http.erl:91]: Sorry, not sure I understand the question :-) right place for what?

File: src/elli_http.erl (hunk 5 above)
Comment (deadtrickster) [reply src/elli_http.erl:286]: No Idea, Content-length was set before me, so I added body

File: src/elli_http.erl (hunk 2 above)
Comment (yurrriq) [reply src/elli_http.erl:91]: This changes moves the line:

``` erlang
{Method, RawPath, V, B0} = get_request(S, PrevB, Opts, Callback),
```

... from before to after:

``` erlang
t(request_start),
```

I just want to be sure that was deliberate and that it's a good decision.

File: src/elli_http.erl (hunk 4 above)
Comment (deadtrickster) [reply src/elli_http.erl:283]: Agree, but I decided to not change this, should I now?

Comment (yurrriq) [reply src/elli_http.erl:283]: I think so.

File: src/elli_http.erl (hunk 5 above)
Comment (yurrriq) [reply src/elli_http.erl:286]: Seems fine then.

File: src/elli_http.erl (hunk 2 above)
Comment (deadtrickster) [reply src/elli_http.erl:91]: Absolutely! I decided to count request_start right after connection accepted. This way we can account slow lorries.

**Comment by yurrriq** [comment]
So I think if you can make the HTTP version in the automatic responses reflect `Req#req.version` then this should be good. That, and confirm the change on line 91 is deliberate or change it.

**Comment by deadtrickster** [comment]
The problem with protocol version is that this rescue function isn't always called when version available e.g.
https://github.com/elli-lib/elli/blob/7e9cadfe3728bcada360c606c869d3361faf4aa3/src/elli_http.erl#L324

**Comment by yurrriq** [comment]
Ah, I see... It could even be that a malformed request doesn't identify itself properly, so there exists a case where we _can't_ know the version.. I guess hardcoding 1.1 is fine then.

**Review by yurrriq — APPROVED** [review]

**Comment by yurrriq** [comment]
LGTM then. Shall I merge?

**Comment by deadtrickster** [comment]
Yeah!

**Pull request merged**